contracts/voting/dao-voting-cw4/src/contract.rs

```diff
@@ -24,124 +25,79 @@ pub fn instantiate(
+            }
 
-#[cfg_attr(not(feature = "library"), entry_point)]
-pub fn execute(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    msg: ExecuteMsg,
-) -> Result<Response, ContractError> {
-    match msg {
-        ExecuteMsg::MemberChangedHook { diffs } => {
-            execute_member_changed_hook(deps, env, info, diffs)
-        }
-    }
-}
+            let mut total_weight = Uint128::zero();
+            for member in initial_members.iter() {
+                deps.api.addr_validate(&member.addr)?;
+                if member.weight > 0 {
+                    // This works because query_voting_power_at_height will return 0 on address missing
+                    // from storage, so no need to store anything.
+                    let weight = Uint128::from(member.weight);
+                    total_weight += weight;
+                }
+            }
 
-pub fn execute_member_changed_hook(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    diffs: Vec<cw4::MemberDiff>,
-) -> Result<Response, ContractError> {
-    let group_contract = GROUP_CONTRACT.load(deps.storage)?;
-    if info.sender != group_contract {
-        return Err(ContractError::Unauthorized {});
-    }
+            if total_weight.is_zero() {
+                return Err(ContractError::ZeroTotalWeight {});
+            }
 
-    let total_weight = TOTAL_WEIGHT.load(deps.storage)?;
-    // As difference can be negative we need to keep track of both
-    // In seperate counters to apply at once and prevent underflow
-    let mut positive_difference: Uint128 = Uint128::zero();
-    let mut negative_difference: Uint128 = Uint128::zero();
-    for diff in diffs {
-        let user_address = deps.api.addr_validate(&diff.key)?;
-        let weight = diff.new.unwrap_or_default();
-        let old = diff.old.unwrap_or_default();
-        // Do we need to add to positive difference or negative difference
-        if weight > old {
-            positive_difference += Uint128::from(weight - old);
-        } else {
-            negative_difference += Uint128::from(old - weight);
+            // We need to set ourself as the CW4 admin it is then transferred to the DAO in the reply
+            let msg = WasmMsg::Instantiate {
+                admin: Some(info.sender.to_string()),
+                code_id: cw4_group_code_id,
+                msg: to_binary(&cw4_group::msg::InstantiateMsg {
+                    admin: Some(env.contract.address.to_string()),
+                    members: initial_members,
+                })?,
+                funds: vec![],
+                label: env.contract.address.to_string(),
+            };
+
+            let msg = SubMsg::reply_on_success(msg, INSTANTIATE_GROUP_REPLY_ID);
+
+            Ok(Response::new()
+                .add_attribute("action", "instantiate")
+                .add_submessage(msg))
         }
+        GroupContract::Existing { address } => {
+            let group_contract = deps.api.addr_validate(&address)?;
```

Comment:
could something go wrong here if the address is valid but does not point to a valid `GroupContract`?

Reply:
Yes. @0xekez, any ideas for validation here?

Reply:
Perhaps it would be nice to use a similar pattern as with `UncheckedDepositInfo` and `CheckedDepositInfo` in `dao-voting` module to validate the group contract

Reply:
Whats the conclusion on this? I think some minimal verification is best to at least try and catch some obvious user errors